2016/proposals/rai-vivek.md

```diff
@@ -0,0 +1,360 @@
+1  | ‘Linux’ | ‘ghostbook’ | ‘3.19.0-47-generic’ | ‘#53~14.04.1-Ubuntu SMP Mon Jan 18 16:09:14 UTC 2016’ | ‘x86_64’ | ‘x86_64’  | foo@bar.com | ‘Tue Mar 8 00:52:35 2016’ | 0
+
+(TABLE software_version)
+
+id | software | installed | version | log
+-----------------------------------------
+1  | python   | True      | 3.5.1   | ‘’
+1  | git      | True      | 1.9.1   | ‘’
+1  | hg       | False     | ‘’      | ‘’
+1  | pip      | True      | 7.9     | ‘A newer version of pip is available.’
+```
+
+Complex queries can be performed by `JOINING` the two tables and querying for
+corresponding field of interest.
+
+#### Concerns
+Both tables share the unique identity given to a particular report.  There is
+a possibility that same user may submit a report twice possibly because there
+were some errors or something else causing duplication.
+
+There might be a couple of ways to tackle this:
+
+1. One way to avoid it would be to generate a unique `machine-id` and use
+   that as a primary identifier key. This key will act as foreign key for key
+   for other table and maintain consistency.
+2. Just execute a SELECT statement on table `machine-info` and see if that
+   information is already record. If yes, then update the records in
+   `software_version` table. Essentially, all the attributes of `machine-info`
+   table may constitute a `UNIQUE_KEY` constraint.
+
+### User interface
+
+* An easily navigable HTML report of diagnostic data.
+
+* Add simple in-browser visualizations.
+  > These visualizations can be served via backend using Python library like
+  > Bokeh or rendered on the client using D3 (or equivalent) library. I prefer the
+  > latter approach.
+
+### Testing
+
+The client-side needs to work reliably across different systems, mainly Linux,
+Windows and OSX machines.
+
+By default, I will develop and test the scripts on my Linux system. For Windows
+and OSX, however, I will have to take help of mentors/community members or my
+friends.
+
+### Updating installation scripts
+
+Modify the installation-test scripts to generate a diagnostic report containing
+machine information and status report as described before. The script should:
+
+* catch abrupt halts (`SIGINT`, for example) and report it to the server.
+* should create a local copy of report log if submission fails.
+
+## Schedule of Deliverables
+
+### April 22nd - May 24nd
+
+Community bonding period. Communicate with members of Software Carpentry
+organization and learn about workshop organization, lesson preparation and other
+details.
+
+The close interaction will help me prepare necessary agenda for a Software
+Carpentry workshop that I am planning to organize locally (India) and eventually
+start an India chapter.
+
+### May 25th -  June 7th
+
+Set up repository. Automated coverage and integration test building using
+[coveralls](https://coveralls.io) and [tavis-ci](https://travis-ci.org/). Will
+follow a test-driven development style.
```

Comment:
We also might have time to add tests to the installation-testing script (if there aren't any).